### ISSUE TITLE
unicorn/no-fn-reference-in-iterator mistaking .find method?

### ISSUE THREAD

**Issue by RobertoMachorro** [sindresorhus/eslint-plugin-unicorn#813]
I'm using a third-party API that contains a method called find. It is not find as in the collection find, but rather their own find. It expects a scalar value (id) and then does its thing.

XO is throwing the following error:

```
✖   82:41  Do not pass function directly to .find(…).  unicorn/no-fn-reference-in-iterator
```

Given that the parameter is scalar, how could this happen? Sample code:

```javascript
const clientId = 20
const client = await oidc.Client.find(clientId)
```

If this is still a valid application of the rule, what would the correct code look like? The [rule doc](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-fn-reference-in-iterator.md) explains the code could be wrapped in an arrow function, which does pass the text. But, is this right?

```javascript
const clientId = 20
const client = await oidc.Client.find(() => clientId)
```

**Comment by RobertoMachorro** [comment]
Follow up, since the arrow function won't actually work with my code, I've disabled the line:

```javascript
// eslint-disable-line unicorn/no-fn-reference-in-iterator  
```

This is the only eslint-disable I have in the code, hoping for some solution!

**Comment by sindresorhus** [comment]
I think we can exclude any `await`'ed expression.

**Comment by papb** [comment]
@RobertoMachorro What you should do is:

```js
const client = await oidc.Client.find(() => clientId())
```

You forgot the last `()`.

That said, I agree with Sindre that the rule should be improved to automatically avoid this false-positive.

**Comment by RobertoMachorro** [comment]
> @RobertoMachorro If you don't want to use `eslint-disable`, you could do this:

Thank you! I'll go with your suggestion for now, and will switch back when this is updated in _unicorn_

> That said, I agree with Sindre that the rule should be improved to automatically avoid this false-positive.

Thanks to you and @sindresorhus for the quick response!

**Comment by fisker** [comment]
I'll work on excluding `await`

**Comment by fisker** [comment]
One second, what if

```js
const promises = [p1, p2, p3];

const isP2 = p => p === p2;

const foo = await promises.find(isP2)
```

?

**Comment by papb** [comment]
@fisker Hmmm, nice catch, in that case ignoring an AwaitExpression would cause a false negative (i.e. no lint error when there should be one).

Suggestions:

* Option 1: do not ignore AwaitExpression for `find`, `reduce` and `reduceRight`, since they are the only ones that could ever return a promise. Proceed to ignore in all other iterators.
* Option 2: accept that these false negatives may happen for `find`, `reduce` and `reduceRight`, considering this a small price to pay for the benefit of having less false positives (?)

I vote for option 1 :)

**Comment by papb** [comment]
@RobertoMachorro Pinging you to call attention to [the edit I made in my previous comment](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/813#issuecomment-684109445). Sorry about that once again.

**Comment by RobertoMachorro** [comment]
> @RobertoMachorro Pinging you to call attention to [the edit I made in my previous comment](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/813#issuecomment-684109445). Sorry about that once again.

No worries, thank you!

**Comment by fisker** [comment]
@papb 

I don't know, ignore other methods (except `.find`, `.reduce`) seems not very useful. 
Is there any lib use something like `await foo.map(bar)`?

**Comment by papb** [comment]
@fisker None that I know of, really... But to me it does seem useful, since I may not know one but it might exist. Also if option 3 is chosen, at least `.find` would be ignored and OP's problem would be solved. @sindresorhus what do you think?

**Comment by sindresorhus** [comment]
I would go for option 3.

**Comment by delucca** [comment]
Same issue here with MirageJS find method.

**Comment by Yohandah** [comment]
I have a class with a .find() method. Unicorn fire an error and thinks I'm using the .find() JS method on iterables ...

**Comment by nicolas-goudry** [comment]
Using [`aigle`](https://github.com/suguru03/aigle) promise library here, and got this error with following code:

```js
Aigle.map(someArray, async (element) => {})

// Do not pass function products directly to .map(…).
```

> I guess it would also throw with `bluebird` or `lodash`, for example

Furthermore, I don't understand usecase for this rule? What’s wrong about defining a function and then use it with `map`, `reduce`, `filter`, etc?

**Comment by papb** [comment]
@nicolas-goudry It is explained [here](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-array-callback-reference.md)
